Add symlinking ability to lbuild

I'm developing library.
I'd like to have ability to symlink files instead of copying for debug. Something in lines of:
```python
def build(env):
    env.outbasepath = "some/dir/to/output"
    env.link("some_file.hpp")
```
I can achieve this behaviour using the following script, but it's way too clunky:
```python
def build(env):
    env.outbasepath = "some/dir/to/output"
    for fn in ["some_file.hpp"]:
        ofn = env.real_outpath(fn)
        env.copy(fn)
        os.unlink(ofn)
        os.symlink(env.localpath(fn), ofn)
```

I tried enabling symlinking in my `modm` project.
Well, it didn't went too well:
```
Compiling C++·· build/release/modm/src/modm/io/iostream.o
<REDACTED>/modm/src/modm/io/iostream.cpp:16:10: fatal error: iostream.hpp: No such file or directory
   16 | #include "iostream.hpp"
      |          ^~~~~~~~~~~~~~
compilation terminated.
```

--- a/lbuild/api.py
+++ b/lbuild/api.py
@@ -121,7 +121,7 @@ class Builder:
         self.parser.validate_modules(build_modules, complete)
         return (build_modules, CallCounter.levels)
 
-    def build(self, outpath, modules=None, simulate=False):
+    def build(self, outpath, modules=None, simulate=False, use_symlinks=False):
         """
         Build the given set of modules.
 
@@ -135,5 +135,6 @@ class Builder:
         """
         build_modules = self._filter_modules(modules)
         buildlog = BuildLog(outpath)
+        lbuild.environment.SYMLINK_ON_COPY = use_symlinks
         self.parser.build_modules(build_modules, buildlog, simulate=simulate)
         return buildlog

--- a/lbuild/environment.py
+++ b/lbuild/environment.py
@@ -25,25 +25,43 @@ import lbuild.exception as le
 from .parser import Parser
 
 SIMULATE = False
+SYMLINK_ON_COPY = False
+
+def default_fn_copy(src, dst):
+    if os.path.lexists(dst):
+        os.remove(dst)
+    if SYMLINK_ON_COPY:
+        os.symlink(src, dst)
+    else:
+        shutil.copy2(src, dst)
 
 
-def _copyfile(sourcepath, destpath, fn_copy=shutil.copy2):
+def _copyfile(sourcepath, destpath, fn_copy=None):
     """
     Copy a file if the source file time stamp is newer than the destination
     timestamp.
     """
+    if fn_copy is None:
+        fn_copy = default_fn_copy
     if not SIMULATE:
         fn_copy(sourcepath, destpath)
 
 
 def _copytree(logger, src, dst, ignore=None,
-              fn_listdir=os.listdir,
-              fn_isdir=os.path.isdir,
-              fn_copy=shutil.copy2):
+              fn_listdir=None,
+              fn_isdir=None,
+              fn_copy=None):
     """
     Implementation of shutil.copytree that overwrites files instead
     of aborting.
     """
+    if fn_listdir is None:
+        fn_listdir = os.listdir
+    if fn_isdir is None:
+        fn_isdir = os.path.isdir
+    if fn_copy is None:
+        fn_copy = default_fn_copy
+
     files = fn_listdir(src)
     if ignore is not None:
         ignored = ignore(src, files)
@@ -307,7 +325,7 @@ class Environment:
         else:
             relative = os.path.join(self.__outpath, relative)
         path = os.path.join(self.__outpath, path)
-        return os.path.relpath(os.path.realpath(path), os.path.realpath(relative))
+        return os.path.relpath(os.path.abspath(path), os.path.abspath(relative))
 
     def generated_local_files(self, filterfunc=None):
         """

--- a/lbuild/main.py
+++ b/lbuild/main.py
@@ -22,7 +22,7 @@ from lbuild.format import format_option_short_description
 
 from lbuild.api import Builder
 
-__version__ = '1.12.4'
+__version__ = '1.12.5'
 
 
 class InitAction:
@@ -297,6 +297,13 @@ class BuildAction(ManipulationActionBase):
             default=False,
             help="Build, but do not write any files. Prints out all generated file names.")
         parser.add_argument(
+            "--symlink",
+            dest="symlink",
+            action="store_true",
+            default=False,
+            help="Use symlinks for file copies. Generated files cannot be "
+                 "copied, so be CAREFUL what files you edit!")
+        parser.add_argument(
             "--no-log",
             dest="buildlog",
             action="store_false",
@@ -308,7 +315,8 @@ class BuildAction(ManipulationActionBase):
 
     @staticmethod
     def perform(args, builder):
-        buildlog = builder.build(args.path, args.modules, simulate=args.simulate)
+        buildlog = builder.build(args.path, args.modules, simulate=args.simulate,
+                                 use_symlinks=args.symlink)
 
         if args.simulate:
             ostream = []

